soccer/planning/RRTPlanner.cpp

```diff
@@ -1,52 +1,136 @@
+#include "RRTPlanner.hpp"
+#include <Constants.hpp>
+#include <Utils.hpp>
+#include <protobuf/LogFrame.pb.h>
+#include "motion/TrapezoidalMotion.hpp"
 
 #include <stdio.h>
 #include <stdlib.h>
 #include <iostream>
 #include <algorithm>
-#include <Constants.hpp>
-#include <Utils.hpp>
 #include <Eigen/Dense>
-#include <protobuf/LogFrame.pb.h>
-
-#include "RRTPlanner.hpp"
-#include "motion/TrapezoidalMotion.hpp"
 
 using namespace std;
-using namespace Planning;
+using namespace Eigen;
+
+namespace Planning {
 
-Geometry2d::Point Planning::randomPoint() {
-    float x =
-        Field_Dimensions::Current_Dimensions.FloorWidth() * (drand48() - 0.5f);
-    float y = Field_Dimensions::Current_Dimensions.FloorLength() * drand48() -
-              Field_Dimensions::Current_Dimensions.Border();
+Geometry2d::Point randomPoint() {
+    const auto& dims = Field_Dimensions::Current_Dimensions;
+    float x = dims.FloorWidth() * (drand48() - 0.5f);
+    float y = dims.FloorLength() * drand48() - dims.Border();
 
     return Geometry2d::Point(x, y);
 }
 
 RRTPlanner::RRTPlanner(int maxIterations) : _maxIterations(maxIterations) {}
 
+bool RRTPlanner::shouldReplan(MotionInstant start, MotionInstant goal,
+                              const MotionConstraints& motionConstraints,
+                              const Geometry2d::ShapeSet* obstacles,
+                              const Path* prevPath) const {
+    if (!prevPath || !prevPath->valid()) return true;
+
+    // if this number of microseconds passes since our last path plan, we
+    // automatically replan
+    const Time kPathExpirationInterval = replanTimeout() * SecsToTimestamp;
+    if ((timestamp() - prevPath->startTime()) > kPathExpirationInterval) {
+        return true;
+    }
+
+    float timeIntoPath =
+        ((float)(timestamp() - prevPath->startTime())) * TimestampToSecs +
+        1.0f / 60.0f;
+
+    MotionInstant target;
+    boost::optional<MotionInstant> optTarget = prevPath->evaluate(timeIntoPath);
+    if (optTarget) {
+        target = *optTarget;
+    } else {
+        // We went off the end of the path, so use the end for calculations.
+        target = *prevPath->destination();
```

Comment:
The check at the top for path validity ensures that it has a destination.  I've changed that to check that destination is non-null now that valid() is gone.